esupy/context_secondary.py

```diff
@@ -0,0 +1,185 @@
+                  uac_url[year])
+        return None
+
+    gdf = crs_harmonize(gdf)  # convert to WGS84
+
+    # check for empty/na geom values
+    ena = sum(gdf['geometry'].is_empty | gdf['geometry'].isna())
+    if ena > 0:
+        log.info(f"{year} Census SHP has {ena} empty + NA geometries")
+    return gdf
+
+def parse_pt_data(df):
+    """
+    Convert df containing "Latitude" and "Longitude" columns to
+    gpd.GeoDataFrame with geometry column of Points
+    :param df: pandas dataframe
+    """
+    # check for na lat/long values
+    pt_na = sum(df['Latitude'].isna() | df['Longitude'].isna())
+    if pt_na != 0:
+        log.info(f"Point data contains {pt_na} nan Lat and/or Long values")
+    gdf = gpd.GeoDataFrame(
+        df, geometry=gpd.points_from_xy(df.Longitude, df.Latitude))
+    gdf = crs_harmonize(gdf)
+    return gdf
+
+def multipoly_agg(gdf):
+    """
+    Aggregate geometry col of polygons and/or multipolygons (no other types)
+    to a single, unified multipolygon collection object
+    :param gdf: GeoPandas geodataframe object
+    """
+    gdf_mp = gdf[gdf['geometry'].type == 'MultiPolygon'] # important to test this piece on urb_2010
+    gdf_p = gdf[gdf['geometry'].type == 'Polygon']
+    if len(gdf) > (len(gdf_mp) + len(gdf_p)):
+        log.error('GeoDataFrame invalid; contains non-polygon geometries')
+        return None
+    # multipolygons are collections of polygons; extract & concatenate into list
+    mp = [poly for multipoly in gdf_mp['geometry'] for poly in multipoly]
+    p = list(gdf_p['geometry']) # concatenate single polygons into a list
+    multipoly = sh.geometry.MultiPolygon(mp + p)  # join lists & convert to single mp
+    return multipoly
+
+def crs_harmonize(gdf):
+    """
+    Convert coordinate reference systems (CRS) to WGS84 (EPSG:4326),
+    or if missing, assume WGS84 and assign
+    :param gdf: GeoPandas geodataframe object
+    """
+    WGS84 = "EPSG:4326"  # EPSG code
+    if gdf.crs is None:  # if crs attribute is empty
+        gdf = gdf.set_crs(WGS84)
+    else:
+        gdf = gdf.to_crs(WGS84)
+    return gdf
+
+def main(df, year, *cmpts):
+    """
+    Handler function to flexibly assign release height ('rh') and/or urban/rural
+    (via 'urb', which initiates geospatial dependencies check) secondary compartments.
+    :param df: pd.DataFrame
+    :param year: int, data year
+    :param cmpts: str, flag(s) for compartment assignment
+    """
+    if 'urb' not in cmpts and 'rh' not in cmpts:
+        log.error('Please pass one or more valid *cmpts string codes: {urb, rh}')
+        return df
+    elif 'urb' in cmpts and not has_geo_pkgs:
+        log.info('Geospatial dependencies of esupy.context_secondary missing; '
+                  'unable to assign urban/rural compartment.\n See esupy README.md.')
+        return df
+    if 'urb' in cmpts:
+        df = urb_intersect(df, year)
+    if 'rh' in cmpts:
+        df = classify_height(df)
```

Comment:
would it make sense to move `classify_height()` up prior to checking `has_geo_pkgs`? That way stack height can still be assigned even if geospatial packages are not available.

Reply:
These separate if-statements can already handle assigning urb, rh, or both based on the contents of the `*cmpts` tuple. If `has_geo_pkgs` = False and `"rh' in cmpts` is True, then `classify_height()` is still called.

Reply:
But in the case of NEI where both `urb` and `rh` are passed, if the packages are not installed line 169 will return the unaltered dataframe

Reply:
Ah yes okay, I'll go ahead and drop 169 and add `and has_geo_pkgs` to 170